src/bms/parse/validity.rs

```diff
@@ -0,0 +1,529 @@
+                }
+                i = j;
+            }
+
+            // Overlap landmine vs single at the same time
+            let mut i = 0;
+            while i < lane_objs.len() {
+                let time = lane_objs[i].offset;
+                let mut j = i;
+                let mut has_visible = false;
+                let mut has_landmine = false;
+                while j < lane_objs.len() && lane_objs[j].offset == time {
+                    match lane_objs[j].kind {
+                        NoteKind::Visible => has_visible = true,
+                        NoteKind::Landmine => has_landmine = true,
+                        _ => {}
+                    }
+                    j += 1;
+                }
+                if has_visible && has_landmine {
+                    invalid.push(ValidityInvalid::OverlapLandmineWithSingle { side, key, time });
+                }
+                i = j;
+            }
+
+            // Helper: check if a time is within [s, e]
+            let time_overlaps_any_ln = |t: ObjTime| -> Option<(ObjTime, ObjTime)> {
+                // Early return if no long notes exist
+                if long_times.is_empty() {
+                    return None;
+                }
+                // Use binary search on sorted long_times to find the first end time >= t
+                let end_i = long_times.partition_point(|&end| end < t);
+                // If end_i is odd, greater than 0, and within bounds, we're inside an interval
+                (end_i > 0 && end_i % 2 == 1 && end_i < long_times.len())
+                    .then_some((long_times[end_i - 1], long_times[end_i]))
+            };
+
+            // Overlap single vs long: any visible single inside any LN interval
+            for o in &lane_objs {
+                if o.kind == NoteKind::Visible
+                    && let Some((s, e)) = time_overlaps_any_ln(o.offset)
+                {
+                    invalid.push(ValidityInvalid::OverlapVisibleSingleWithLong {
+                        side,
+                        key,
+                        time: o.offset,
+                        ln_start: s,
+                        ln_end: e,
+                    });
+                }
+            }
+
+            // Landmine vs long: warn once per LN interval, at long start, if any landmine inside interval
+            for (s, e) in &ln_intervals {
+                let mut has_landmine = false;
+                for o in &lane_objs {
+                    if o.kind == NoteKind::Landmine && o.offset >= *s && o.offset <= *e {
+                        has_landmine = true;
+                        break;
+                    }
+                }
+                if has_landmine {
+                    invalid.push(ValidityInvalid::LandmineOverlapsLongAtStart {
+                        side,
+                        key,
+                        ln_start: *s,
+                        ln_end: *e,
+                    });
+                }
+            }
```

Comment:
It can be done same as "Overlap single vs long". Then `ln_intervals` will become needless.
```suggestion
            for o in &lane_objs {
                if o.kind == NoteKind::Landmine
                    && let Some((s, e)) = time_overlaps_any_ln(o.offset)
                    && s == o.offset
                {
                    invalid.push(ValidityInvalid::LandmineOverlapsLongAtStart {
                        side,
                        key,
                        ln_start: s,
                        ln_end: e,
                    });
                }
            }
```